src/broker/cas_ssl.c

```diff
@@ -0,0 +1,295 @@
+      return err_code;
+    }
+
+
+  if ((ssl = SSL_new (ctx)) == NULL)
+    {
+      cas_log_write_and_end (0, true, "SSL: Creating SSL context failed.");
+      SSL_CTX_free (ctx);
+      return ER_SSL_GENERAL;
+    }
+
+  if (SSL_set_fd (ssl, sd) == 0)
+    {
+      cas_log_write_and_end (0, true, "SSL: Cannot associate with socket.");
+      SSL_free (ssl);
+      ssl = NULL;
+      return ER_SSL_GENERAL;
+    }
+
+  err_code = SSL_accept (ssl);
+  if (err_code < 0)
+    {
+      err_code = SSL_get_error (ssl, err_code);
+      err = ERR_get_error ();
+      cas_log_write_and_end (0, true, "SSL: Accept failed - '%s'", ERR_error_string (err, NULL));
+      SSL_free (ssl);
+      ssl = NULL;
+      return ER_SSL_GENERAL;
+    }
+
+#if defined (WINDOWS)
+  argp = SOCKET_NONBLOCK;
+  ioctlsocket (sd, FIONBIO, &argp);
+#else
+  fcntl (sd, F_SETFL, oflags);
+#endif
+
+  ssl_client = true;
+
+  return 0;
+}
+
+int
+cas_ssl_read (int sd, char *buf, int size)
+{
+  int nread;
+
+  if (IS_INVALID_SOCKET (sd) || ssl == NULL)
+    {
+      cas_log_write_and_end (0, true, "SSL: READ attempt for brokern connection");
+      return ER_SSL_GENERAL;
+    }
+
+#if defined(WINDOWS)
+  u_long argp = SOCKET_BLOCK;
+  ioctlsocket (sd, FIONBIO, &argp);
+#else
+  int oflags, flags = fcntl (sd, F_GETFL, 0);
+  oflags = flags;
+  flags = flags & ~O_NONBLOCK;
+  fcntl (sd, F_SETFL, flags);
+#endif
+
+  nread = SSL_read (ssl, buf, size);
+
+#if defined(WINDOWS)
+  argp = SOCKET_NONBLOCK;
+  ioctlsocket (sd, FIONBIO, &argp);
+#else
+  fcntl (sd, F_SETFL, oflags);
+#endif
+  return nread;
+}
+
+int
+cas_ssl_write (int sd, const char *buf, int size)
+{
+  int nwrite;
+
+  if (IS_INVALID_SOCKET (sd) || ssl == NULL)
+    {
+      cas_log_write_and_end (0, true, "SSL: WRITE attempt for brokern connection");
+      return ER_SSL_GENERAL;
+    }
+#if defined(WINDOWS)
+  u_long argp = SOCKET_BLOCK;
+  ioctlsocket (sd, FIONBIO, &argp);
+#else
+  int oflags, flags = fcntl (sd, F_GETFL, 0);
+  oflags = flags;
+  flags = flags & ~O_NONBLOCK;
+  fcntl (sd, F_SETFL, flags);
+#endif
+
+  nwrite = SSL_write (ssl, buf, size);
+
+#if defined(WINDOWS)
+  argp = SOCKET_NONBLOCK;
+  ioctlsocket (sd, FIONBIO, &argp);
+#else
+  fcntl (sd, F_SETFL, oflags);
+#endif
+  return nwrite;
+}
+
+void
+cas_ssl_close (int client_sock_fd)
+{
+  if (ssl)
+    {
+      SSL_free (ssl);
+      ssl = NULL;
+    }
+}
+
+static int
+cas_ssl_validity_check (SSL_CTX * ctx)
+{
+  ASN1_TIME *not_before, *not_after;
+  X509 *crt;
+
+  crt = SSL_CTX_get0_certificate (ctx);
+
+  if (crt == NULL)
+    return ER_SSL_GENERAL;
+
+  not_after = X509_getm_notAfter (crt);
+  if (X509_cmp_time (not_after, NULL) != 1)
+    {
+      return ER_CERT_EXPIRED;
+    }
+
+  not_before = X509_getm_notBefore (crt);
+  if (X509_cmp_time (not_before, NULL) != -1)
+    {
+      return ER_SSL_GENERAL;
+    }
+
+  return 0;
+}
+
+bool
+is_ssl_data_ready (int sock_fd)
+{
+  return (SSL_has_pending (ssl) == 1 ? true : false);
```

Comment:
SSL_has_pending() returns the number of bytes which have been processed, buffered, so I think you should change this code as the following. 

return (SSL_has_pending (ssl) >= 1 ? true : false)

Reply:
SSL_has_pending () returns value 0 or 1.
I think you talking about SSL_pending ()
either 

```
return (SSL_has_pending (ssl) == 1 ? true : false);
or
return (SSL_pending (ssl) >= 1 ? true : false);
```

Reply:
I got it.